### ISSUE TITLE
`isValidUtf8ByteArray` returns false positives

### ISSUE THREAD

**Issue by sol** [haskell/text#540]
Repro:
```
ghci> :set -XOverloadedLists
ghci> import Data.Text.Internal.Validate
ghci> isValidUtf8ByteArray [128] 0 1
False
ghci> isValidUtf8ByteArray [0, 128] 1 1
True
```
(GHC 9.8.1 / Linux x64)

**Comment by sol** [comment]
1. Apparently my GHC provided `text-2.1` is not built with `simdutf`.
2. The Haskell fallback looks dodgy.  I guess the `len` in line 137 should be `end` where `end = off + len` (same in line 143).

https://github.com/haskell/text/blob/73620de89d43ee50de2d15b7bc0843bf6d6e9b9a/src/Data/Text/Internal/Validate.hs#L132-L137
https://github.com/haskell/text/blob/73620de89d43ee50de2d15b7bc0843bf6d6e9b9a/src/Data/Text/Internal/Validate.hs#L143

@andrewthad where did you take this code from?  Are there any tests?

**Comment by Bodigrim** [comment]
> 2\. The Haskell fallback looks dodgy.  I guess the `len` in line 137 should be `end` where `end = off + len` (same in line 143).

Good catch. Could you possibly prepare a PR please?

**Comment by sol** [comment]
> Good catch. Could you possibly prepare a PR please?

I am more inclined to fix this by using `bytestring_is_valid_utf8`.  However, after spending some time on it, I am close to the conclusion that that implementation is dodgy as well.

**Comment by sol** [comment]
https://github.com/haskell/bytestring/issues/620

**Comment by andrewthad** [comment]
@sol The implementation of `isValidUtf8ByteArrayHaskell#` is copied from one of the shims for `Data.Text.Internal.Validate.isValidUtf8ByteString`:

```
isValidUtf8ByteString :: ByteString -> Bool
#ifdef SIMDUTF
isValidUtf8ByteString bs = withBS bs $ \fp len -> unsafeDupablePerformIO $
  unsafeWithForeignPtr fp $ \ptr -> (/= 0) <$> c_is_valid_utf8_ptr_unsafe ptr (fromIntegral len)
#else
#if MIN_VERSION_bytestring(0,11,2)
isValidUtf8ByteString = B.isValidUtf8
#else
isValidUtf8ByteString bs = start 0
  where
    start ix
      | ix >= B.length bs = True
      | otherwise = case utf8DecodeStart (B.unsafeIndex bs ix) of
        Accept{} -> start (ix + 1)
        Reject{} -> False
        Incomplete st _ -> step (ix + 1) st
    step ix st
      | ix >= B.length bs = False
      -- We do not use decoded code point, so passing a dummy value to save an argument.
      | otherwise = case utf8DecodeContinue (B.unsafeIndex bs ix) st (CodePoint 0) of
        Accept{} -> start (ix + 1)
        Reject{} -> False
        Incomplete st' _ -> step (ix + 1) st'
#endif
#endif
```

I've investigated a little to figure out where `isValidUtf8ByteString` came from. Before I added anything, `isValidUtf8ByteString` was already there, but it was named `isValidBS` instead. But, between the time that I put up my PR and the time it was merged, commit 7ef771dc8cfddc1bb84abaec35e58c384e4f04bc made it in, which removed `isValidBS`. So, from commit history on `master`, it looks like `isValidUtf8ByteString` just came out of nowhere, but if you go back to 6f1917df2075d8763efd1beb681cc3503b794573, you can its predecessor `isValidBS`.

I botched the adaptation though, and the comparison should be `ix >= off + len` instead of `ix >= len`. (Or computing `end = off + len` in a `where` clause and then using `ix >= end` instead.

There are not any tests for this function that I'm aware of.

**Comment by Bodigrim** [comment]
> I botched the adaptation though, and the comparison should be `ix >= off + len` instead of `ix >= len`. (Or computing `end = off + len` in a `where` clause and then using `ix >= end` instead.

We need to fix this before the next release, which is soon. Any volunteers?
